support/compressxdr/compress_xdr.go

```diff
@@ -3,36 +3,50 @@ package compressxdr
 import (
 	"io"
 
-	"github.com/stellar/go/support/errors"
+	xdr3 "github.com/stellar/go-xdr/xdr3"
 )
 
-const (
-	GZIP = "gzip"
-)
+func NewXDREncoder(compressor Compressor, xdrPayload interface{}) XDREncoder {
+	return XDREncoder{Compressor: compressor, XdrPayload: xdrPayload}
+}
+
+func NewXDRDecoder(compressor Compressor, xdrPayload interface{}) XDRDecoder {
+	return XDRDecoder{Compressor: compressor, XdrPayload: xdrPayload}
 
-type XDREncoder interface {
-	WriteTo(w io.Writer) (int64, error)
 }
 
-type XDRDecoder interface {
-	ReadFrom(r io.Reader) (int64, error)
-	Unzip(r io.Reader) ([]byte, error)
+// XDREncoder combines compression with XDR encoding
+type XDREncoder struct {
+	Compressor Compressor
+	XdrPayload interface{}
 }
 
-func NewXDREncoder(compressionType string, xdrPayload interface{}) (XDREncoder, error) {
-	switch compressionType {
-	case GZIP:
-		return &XDRGzipEncoder{XdrPayload: xdrPayload}, nil
-	default:
-		return nil, errors.Errorf("invalid compression type %s, not supported", compressionType)
+// WriteTo writes the XDR compressed encoded data
+func (e XDREncoder) WriteTo(w io.Writer) (int64, error) {
+	zw, err := e.Compressor.NewWriter(w)
+	if err != nil {
+		return 0, err
 	}
+	defer zw.Close()
+
+	n, err := xdr3.Marshal(zw, e.XdrPayload)
+	return int64(n), err
 }
 
-func NewXDRDecoder(compressionType string, xdrPayload interface{}) (XDRDecoder, error) {
-	switch compressionType {
-	case GZIP:
-		return &XDRGzipDecoder{XdrPayload: xdrPayload}, nil
-	default:
-		return nil, errors.Errorf("invalid compression type %s, not supported", compressionType)
+// XDRDecoder combines decompression with XDR decoding
+type XDRDecoder struct {
+	Compressor Compressor
+	XdrPayload interface{}
+}
+
+// ReadFrom reads XDR compressed encoded data
+func (d *XDRDecoder) ReadFrom(r io.Reader) (int64, error) {
```

Comment:
```suggestion
func (d XDRDecoder) ReadFrom(r io.Reader) (int64, error) {
```
this doesn't need to be a pointer receiver because it does not modify the XDRDecoder  struct